File: Problem_List/20.Valid_Parentheses/code.cpp
#include <stack>
using namespace std;;

class Solution {
public:
    bool isValid(string s) {
        stack<char> str;
        for(char c : s){
            if(c == '(' || c == '{' || c == '['){
                str.push(c);
            }else{
                if(str.empty())
                    return false;
                    
                if(c == ')'){
                    if(str.top() == '(')
                        str.pop();
                    else
                        return false;
                }
                else if(c == '}'){
                    if(str.top() == '{')
                        str.pop();
                    else
                        return false;
                }
                else if(c == ']')
                    if(str.top() == '[')
                        str.pop();
                    else
                        return false;
            }
        }
        if (str.empty())
            return true;
        else
            return false;
    }
};

File: Problem_List/20.Valid_Parentheses/code.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        mapping = {'{' : '}', '[' : ']', '(' : ')'}

        for char in s:
            if char in mapping:
                stack.append(char)
            else:
                if not stack or char != mapping[stack.pop()]:
                    return False

        return not stack
